include/sys/zstd/zstd.h

```diff
@@ -56,18 +56,13 @@ typedef struct zfs_zstd_header {
 
 	/*
 	 * Version and compression level
-	 * We use a union to be able to big endian encode a single 32 bit
-	 * unsigned integer, but still access the individual bitmasked
-	 * components easily.
+	 * We used to use a union to reference compression level
+	 * and version easily, but as it turns out, relying on the
+	 * ordering of bitfields is not remotely portable.
+	 * So now we have get/set functions in zfs_zstd.c for
+	 * manipulating this in just the right way forever.
 	 */
-	union {
-		uint32_t raw_version_level;
-		struct {
-			uint32_t version : 24;
-			uint8_t level;
-		};
-	};
-
+	uint32_t raw_version_level;
```

Comment:
I wonder if some of your other logic could be simplified by declaring this as `uint8_t[4]`.

Reply:
I believe it would let me drop most of the boilerplate around wherever I do pointer math; I'll experiment with it once I get the macro usage right.

Reply:
With everything I dropped, it didn't benefit anything, so I didn't do it.